pkg/service/legacy/service.go

```diff
@@ -0,0 +1,158 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"github.com/storacha/go-capabilities/pkg/assert"
+	"github.com/storacha/go-ucanto/core/delegation"
+	"github.com/storacha/go-ucanto/core/ipld"
+	"github.com/storacha/go-ucanto/principal"
+	"github.com/storacha/indexing-service/pkg/blobindex"
+	"github.com/storacha/indexing-service/pkg/bytemap"
+	"github.com/storacha/indexing-service/pkg/internal/link"
+	"github.com/storacha/indexing-service/pkg/service/queryresult"
+	"github.com/storacha/indexing-service/pkg/types"
+)
+
+type IndexingService struct {
+	id              principal.Signer
+	indexingService types.Service
+	blockIndexStore BlockIndexStore
+	bucketURL       url.URL
+}
+
+func (l *IndexingService) Cache(ctx context.Context, provider peer.AddrInfo, claim delegation.Delegation) error {
+	return l.indexingService.Cache(ctx, provider, claim)
+}
+
+func (l *IndexingService) Get(ctx context.Context, claim datamodel.Link) (delegation.Delegation, error) {
+	return l.indexingService.Get(ctx, claim)
+}
+
+func (l *IndexingService) Publish(ctx context.Context, claim delegation.Delegation) error {
+	return l.indexingService.Publish(ctx, claim)
+}
+
+func (l *IndexingService) Query(ctx context.Context, q types.Query) (types.QueryResult, error) {
+	results, err := l.indexingService.Query(ctx, q)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(results.Claims()) > 0 || len(results.Indexes()) > 0 || len(q.Hashes) == 0 {
+		return results, nil
+	}
+
+	// lets see if we can materialize some location claims
+	content := assert.FromHash(q.Hashes[0])
+	records, err := l.blockIndexStore.Query(ctx, content.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	var locs []assert.LocationCaveats
+	var derivedLocs []assert.LocationCaveats
+	for _, r := range records {
+		u, err := url.Parse(r.CarPath)
+		if err != nil {
+			// non-URL is legacy region/bucket/key format
+			// e.g. us-west-2/dotstorage-prod-1/raw/bafy...
+			parts := strings.Split(r.CarPath, "/")
+			key := strings.Join(parts[2:], "/")
+			shard, err := bucketKeyToShardLink(key)
+			if err != nil {
+				continue
+			}
+
+			u = l.bucketURL.JoinPath(fmt.Sprintf("/%s/%s.car", shard.String(), shard.String()))
+			derivedLocs = append(derivedLocs, assert.LocationCaveats{
+				Content:  content,
+				Location: []url.URL{*u},
+				Range:    &assert.Range{Offset: r.Offset, Length: &r.Length},
+			})
+			continue
+		}
+
+		locs = append(locs, assert.LocationCaveats{
+			Content:  content,
+			Location: []url.URL{*u},
+			Range:    &assert.Range{Offset: r.Offset, Length: &r.Length},
+		})
+	}
+
+	// prefer items with non derived location URLs
+	if len(locs) == 0 {
+		locs = derivedLocs
+	}
```

Comment:
not sure I get why we need `locs` and `derivedLocs`. `records` are data about blocks associated with some `content`. Can these records have mixed `CarPath` formats? I mean, I'd expect all records to either have `CarPath`s that are URLs or legacy. If they all have the same format, we will have all items in either `locs` or `derivedLocs` and we only need one of them.

From the code, I assume that the block index store may contain data for all the blocks that belong to `content` in both formats, is that correct?

Reply:
It depends, if the same block is was uploaded in a CAR a long time again and then again in a CAR more recently it'll have multiple entries in the table. That said, I think you're righ it doesn't really matter which ones we use - URLs or old paths... so I've removed.